File: contract-agent/src/systemStatus.ts
import os from "node:os";
import { execSync } from "node:child_process";
import { nowIsoKst } from "@slash-agent/contracts";

export interface SystemStatusResult {
  cpuPercent: number;
  memoryPercent: number;
  memoryTotalMb: number;
  memoryUsedMb: number;
  diskPercent: number | null;
  diskTotalMb: number | null;
  diskUsedMb: number | null;
  collectedAt: string;
}

function cpuPercentApprox(): number {
  const load1 = os.loadavg()[0] ?? 0;
  const cores = os.cpus().length || 1;
  return Math.max(0, Math.min(100, Math.round((load1 / cores) * 100)));
}

function diskUsage(): { percent: number; totalMb: number; usedMb: number } | null {
  try {
    const output = execSync("df -k /").toString();
    const lines = output.trim().split("\n");
    const parts = lines[lines.length - 1].trim().split(/\s+/);
    const totalKb = Number(parts[1]);
    const usedKb = Number(parts[2]);
    if (!totalKb) return null;
    return {
      percent: Math.round((usedKb / totalKb) * 100),
      totalMb: Math.round(totalKb / 1024),
      usedMb: Math.round(usedKb / 1024),
    };
  } catch {
    return null;
  }
}

/** 이 macOS 실기기의 실제 CPU/메모리/디스크 값을 수집한다 (Gemma/AWS 같은 실 서비스 흉내 없음). */
export function collectSystemStatus(): SystemStatusResult {
  const totalMem = os.totalmem();
  const freeMem = os.freemem();
  const usedMem = totalMem - freeMem;
  const disk = diskUsage();
  return {
    cpuPercent: cpuPercentApprox(),
    memoryPercent: Math.round((usedMem / totalMem) * 100),
    memoryTotalMb: Math.round(totalMem / 1024 / 1024),
    memoryUsedMb: Math.round(usedMem / 1024 / 1024),
    diskPercent: disk?.percent ?? null,
    diskTotalMb: disk?.totalMb ?? null,
    diskUsedMb: disk?.usedMb ?? null,
    collectedAt: nowIsoKst(),
  };
}

File: slash-python-pc-runner/src/slash_pc_runner/system_status.py
"""SYSTEM_STATUS 작업 실행 — systemStatus.ts 대응. 이 실기기의 CPU/메모리/디스크 실측값."""

from __future__ import annotations

import shutil

import psutil

from .protocol import now_iso_kst


def collect_system_status() -> dict:
    vm = psutil.virtual_memory()
    try:
        disk = shutil.disk_usage("/")
        disk_percent = round(disk.used / disk.total * 100)
        disk_total_mb = disk.total // (1024 * 1024)
        disk_used_mb = disk.used // (1024 * 1024)
    except OSError:
        disk_percent = disk_total_mb = disk_used_mb = None

    return {
        "cpuPercent": psutil.cpu_percent(interval=0.1),
        "memoryPercent": round(vm.percent),
        "memoryTotalMb": vm.total // (1024 * 1024),
        "memoryUsedMb": (vm.total - vm.available) // (1024 * 1024),
        "diskPercent": disk_percent,
        "diskTotalMb": disk_total_mb,
        "diskUsedMb": disk_used_mb,
        "collectedAt": now_iso_kst(),
    }
